File: services/core-api/app/schemas/story_context.py
"""Schemas for the story context REST endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel


FactCategory = Literal[
    "person", "place", "date", "event", "emotion", "relationship", "object"
]
FactSource = Literal["story", "conversation"]
FactStatus = Literal["active", "pinned", "dismissed"]


class ContextFactResponse(BaseModel):
    """A single extracted fact."""

    id: UUID
    category: FactCategory
    content: str
    detail: str | None
    source: FactSource
    status: FactStatus
    created_at: datetime

    model_config = {"from_attributes": True}


class StoryContextResponse(BaseModel):
    """Full context for a story: summary + facts."""

    id: UUID
    story_id: UUID
    summary: str | None
    summary_updated_at: datetime | None
    extracting: bool
    facts: list[ContextFactResponse]

    model_config = {"from_attributes": True}


class ExtractRequest(BaseModel):
    """Request to trigger context extraction."""

    force: bool = False


class ExtractResponse(BaseModel):
    """Response from extraction trigger."""

    status: Literal["extracting", "cached"]


class FactStatusUpdate(BaseModel):
    """Request to update a fact's status."""

    status: FactStatus


class PinnedFact(BaseModel):
    """A pinned fact sent to the rewrite endpoint."""

    category: FactCategory
    content: str
    detail: str | None = None

File: apps/web/src/features/evolve-workspace/api/storyContext.ts
// apps/web/src/features/evolve-workspace/api/storyContext.ts
import { apiGet, apiPost, apiPatch } from '@/lib/api/client';

// --- Types ---

export type FactCategory =
  | 'person'
  | 'place'
  | 'date'
  | 'event'
  | 'emotion'
  | 'relationship'
  | 'object';

export type FactSource = 'story' | 'conversation';
export type FactStatus = 'active' | 'pinned' | 'dismissed';

export interface ContextFact {
  id: string;
  category: FactCategory;
  content: string;
  detail: string | null;
  source: FactSource;
  status: FactStatus;
  created_at: string;
}

export interface StoryContextResponse {
  id: string;
  story_id: string;
  summary: string | null;
  summary_updated_at: string | null;
  extracting: boolean;
  facts: ContextFact[];
}

export interface ExtractResponse {
  status: 'extracting' | 'cached';
}

// --- API Functions ---

export async function getStoryContext(
  storyId: string,
): Promise<StoryContextResponse> {
  return apiGet<StoryContextResponse>(`/api/stories/${storyId}/context`);
}

export async function extractContext(
  storyId: string,
  force = false,
): Promise<ExtractResponse> {
  return apiPost<ExtractResponse>(`/api/stories/${storyId}/context/extract`, {
    force,
  });
}

export async function updateFactStatus(
  storyId: string,
  factId: string,
  status: FactStatus,
): Promise<ContextFact> {
  return apiPatch<ContextFact>(
    `/api/stories/${storyId}/context/facts/${factId}`,
    { status },
  );
}
